core/benches/bls_sigverify.rs

```diff
@@ -1,172 +1,260 @@
+            rank: i as u16,
+        });
+    }
+
+    let mut builder = VoteCertificateBuilder::new(certificate);
+    builder.aggregate(&all_vote_messages).unwrap();
+    let cert_message = builder.build().unwrap();
+    ConsensusMessage::Certificate(cert_message)
+}
+
+// Scenario 1: One batch with two votes.
+fn generate_two_votes_batch(env: &BenchEnvironment) -> Vec<PacketBatch> {
+    // Use Notarization votes as in the original hardcoded data structure.
+    let vote = Vote::new_notarization_vote(BENCH_SLOT, Hash::new_unique());
+    let payload = bincode::serialize(&vote).unwrap();
+
+    // Vote 1 (Signed by Rank 0)
+    let kp1 = &env.validator_keypairs[0].bls_keypair;
+    let sig1: BlsSignature = kp1.sign(&payload).into();
+    let msg1 = ConsensusMessage::Vote(VoteMessage {
+        vote,
+        signature: sig1,
+        rank: 0,
+    });
+
+    // Vote 2 (Signed by Rank 1)
+    let kp2 = &env.validator_keypairs[1].bls_keypair;
+    let sig2: BlsSignature = kp2.sign(&payload).into();
+    let msg2 = ConsensusMessage::Vote(VoteMessage {
+        vote,
+        signature: sig2,
+        rank: 1,
+    });
+
+    let packets = vec![message_to_packet(&msg1), message_to_packet(&msg2)];
+    vec![PinnedPacketBatch::new(packets).into()]
+}
+
+// Scenario 2: One batch with a single vote.
+fn generate_single_vote_batch(env: &BenchEnvironment) -> Vec<PacketBatch> {
+    // Use a Finalization vote as in the original hardcoded data structure.
+    let vote = Vote::new_finalization_vote(BENCH_SLOT);
+    let payload = bincode::serialize(&vote).unwrap();
+
+    // Vote 1 (Signed by Rank 0)
+    let kp = &env.validator_keypairs[0].bls_keypair;
+    let sig: BlsSignature = kp.sign(&payload).into();
+    let msg = ConsensusMessage::Vote(VoteMessage {
+        vote,
+        signature: sig,
+        rank: 0,
     });
 
-    stage.join().unwrap();
+    let packets = vec![message_to_packet(&msg)];
+    vec![PinnedPacketBatch::new(packets).into()]
+}
+
+// Scenario 3: A batch with a single certificate.
+fn generate_single_cert_batch(env: &BenchEnvironment) -> Vec<PacketBatch> {
+    let hash = Hash::new_unique();
+    // Generate a Base2 certificate
+    let msg = create_base2_cert_message(env, BENCH_SLOT, hash);
+    let packets = vec![message_to_packet(&msg)];
+    vec![PinnedPacketBatch::new(packets).into()]
+}
+
+// Scenario 4: A batch with two certificates (one Base2, one Base3).
+fn generate_two_certs_batch(env: &BenchEnvironment) -> Vec<PacketBatch> {
+    let hash1 = Hash::new_unique();
+    let hash2 = Hash::new_unique();
+
+    // Cert 1 (Base2 - Notarize)
+    let msg1 = create_base2_cert_message(env, BENCH_SLOT, hash1);
+    // Cert 2 (Base3 - NotarizeFallback)
+    let msg2 = create_base3_cert_message(env, BENCH_SLOT + 1, hash2);
+
+    let packets = vec![message_to_packet(&msg1), message_to_packet(&msg2)];
+    vec![PinnedPacketBatch::new(packets).into()]
 }
 
-fn bench_bls(c: &mut Criterion) {
+fn bench_votes(c: &mut Criterion) {
```

Comment:
nit: maybe we can put some rough numbers here how often we see these two scenarios in the log

Reply:
I commented on the distribution of vote batches. 

For certs, I think we need a slightly more sophisticated bench in a follow-up. Pretty much all certs are identical certs, but the numbers in a batch can vary (a single cert in a batch, all the way up to 6 identical certs in a batch). Currently, even if we have 6 identical certs in a single batch (but with different sigs), we do verify each of them.

In a follow-up, I'll address this and then add a more sophisticated bench logic for this case.